crates/ra_hir_def/src/db.rs

```diff
@@ -90,6 +91,9 @@ pub trait DefDatabase: InternDatabase + AstDatabase {
     #[salsa::invoke(Attrs::attrs_query)]
     fn attrs(&self, def: AttrDefId) -> Attrs;
 
+    #[salsa::invoke(Visibility::visibility_query)]
+    fn visibility(&self, def: VisibilityDefId) -> Visibility;
```

Comment:
Hm, I am not sure if we should really make this a query... Seems like it would be easier and more effective to just compute visibility directly in various `x_data` queries? Especially if we store an unresolved `ModPath` inside visibility.

Reply:
Hm, I wonder if that logic applies to generics as well... It seems reasonable to do all "data" lowering in once step, and then implement various traits on `XData` structs, rather than on IDs. (Take this with a grain of salt, as I am not exactly awake at the moment :)

Reply:
After working on integrating visibility in name resolution, I agree -- since visibility for items needs to be handled in the collector, this query would only really be used for methods and fields anyway.

Reply:
Hm, I now see that there's another problem here: we duplicate info between `raw` and `x_data` queries. 

For example, we extract `Name` out of `ast::StructDef` both in `RawItemsCollector::add_item` and in `struct_data_query`. Naturally, the two places do it differently: the former ignores anonymous structs, while the latter uses `Name::missing` (which I think is a dead code). 

But that probably is an unrelated problem. For know, I think it's ok to duplicate visbility as well between raw and data queries.

Reply:
I think in the case of visibility, we probably don't even need it in the item data. It makes more sense as a property of a specific (imported or defined) name as opposed to the item itself anyway, i.e. you shouldn't ask "is this item visible", but "is this path visible".